corehq/apps/sso/async_handlers.py

```diff
@@ -0,0 +1,179 @@
+    UserExemptFromSingleSignOn,
+)
+
+
+class Select2IdentityProviderHandler(BaseSelect2AsyncHandler):
+    slug = 'select2_identity_provider'
+    allowed_actions = [
+        'owner',
+    ]
+
+    @property
+    def owner_response(self):
+        accounts = BillingAccount.objects.filter(is_customer_billing_account=True)
+        if self.search_string:
+            accounts = accounts.filter(name__icontains=self.search_string)
+        return [(a.id, a.name) for a in accounts.order_by('name')]
+
+
+class BaseLinkedObjectAsyncHandler(BaseAsyncHandler):
+    allowed_actions = [
+        'get_linked_objects',
+        'add_object',
+        'remove_object',
+    ]
+
+    def get_linked_objects(self):
+        raise NotImplementedError("please implement get_linked_objects")
+
+    def add_object(self):
+        raise NotImplementedError("please implement add_object")
+
+    def remove_object(self):
+        raise NotImplementedError("please implement add_object")
+
+    @property
+    def get_linked_objects_response(self):
+        return {
+            'linkedObjects': self.get_linked_objects(),
+        }
+
+    @property
+    def add_object_response(self):
+        self.add_object()
+        return {
+            'linkedObjects': self.get_linked_objects(),
+        }
+
+    @property
+    def remove_object_response(self):
+        self.remove_object()
+        return {
+            'linkedObjects': self.get_linked_objects(),
+        }
+
+
+class IdentityProviderAdminAsyncHandler(BaseLinkedObjectAsyncHandler):
+    slug = 'identity_provider_admin'
+
+    def get_linked_objects(self):
+        return list(AuthenticatedEmailDomain.objects.filter(
+            identity_provider__slug=self.idp_slug
+        ).order_by('email_domain').values_list('email_domain', flat=True))
+
+    def add_object(self):
+        if AuthenticatedEmailDomain.objects.filter(email_domain=self.email_domain).exists():
+            raise AsyncHandlerError(
+                f"Email domain {self.email_domain} is already associated"
+                f"with an identity provider."
+            )
+        AuthenticatedEmailDomain.objects.create(
+            identity_provider=self.identity_provider,
+            email_domain=self.email_domain,
+        )
+
+    def remove_object(self):
+        existing_email_domain = AuthenticatedEmailDomain.objects.filter(
+            email_domain=self.email_domain
+        )
+        if not existing_email_domain.exists():
+            raise AsyncHandlerError(
+                f"No email domain exists with the name {self.email_domain}."
+            )
+        existing_email_domain.delete()
+
+    @property
+    @memoized
+    def idp_slug(self):
+        return self.data.get('requestContext[idpSlug]')
+
+    @property
+    @memoized
+    def identity_provider(self):
+        return IdentityProvider.objects.get(slug=self.idp_slug)
+
+    @property
+    @memoized
+    def email_domain(self):
+        return self.data.get('objectName')
```

Comment:
easier to read, setup, and subclass imo. this is the pattern we've followed for class-based views

Reply:
also, it wouldn't work to call it in `__init__` actually. see
https://stackoverflow.com/questions/46599145/django-class-view-init

The common recommendation is actually to NEVER start from `__init__` with class-based views.

Reply:
Thanks for that link -- I've been looking for a reason to dig deeper into class-based views, and this let me get into the nitty-gritty. I'm guessing this pattern is a historical artifact, caused by having no logical place to initialize properties. The article you linked suggested overriding `dispatch()`, and that used to be the main way to do it. I'd imagine we went for memoized properties because it was slightly more intuitive than overriding dispatch. The correct way now, it seems, is to [override `setup()`](https://docs.djangoproject.com/en/3.1/ref/class-based-views/base/#django.views.generic.base.View.setup). However, that was only [introduced in 2.2](https://docs.djangoproject.com/en/3.1/releases/2.2/#generic-views). I get why we did it in the past, but I have a hard time advocating for solutions just because that's how it was done previously.